### PR TITLE
[MIRROR] Speeds up the preference menu, significantly. Adds object pooling, other stuff too [MDB IGNORE]

### PR DESCRIPTION
**Pull request by SkyratBot**
Original PR: https://github.com/tgstation/tgstation/pull/63225
--------------------

## About The Pull Request

Hello again bestie.
I'm very miffed about how slow the preferences menu is. Especially in light of #63118
It's very slow, and because it runs off Topic (which is technically a verb) it runs after even sendmaps. Lots of lagspikes.

Sooooo I'm going to optimize it. Need more user experience time for atmos after all.
Let's break down where it slows down/speeds up yeah?

Oh and before I start, all profiles are setup like this.
I intercept the ui_act args intended for preferences, and store them in a list on SSInput. Every time SSInput fires, we call the ui_act once. Preferences handily doesn't cache ui acts, so this is very effective.
Anyway I call the ui_act 1.2k times, and then stop. Hopefully this is a large enough call count to make any issues very clear from the profiler.
 
 Using my ethereal static for this, if it matters to you
image

To start with, here's a profile of updating the preferences as it is on master, no outfit, just bare naked
[Pre Optimization No Gear](https://github.com/tgstation/tgstation/files/7647114/pre.optimization.no.gear.csv)
Takes roughly 30 seconds to run 1.2k calls. Not great, roughly 25 milliseconds

Ok so how long does it take to do the same thing, but to a character wearing a CEs outfit?
[Pre Optimization CE](https://github.com/tgstation/tgstation/files/7647132/pre.optimization.CE.csv)
141 SECONDS. Jesus christ, that's roughly 117 miliseconds, almost a 10th of a second. No fucking wonder this lags sometimes.

Ok so why then. Well the answer is a bit longwinded, but let's start with the start yeah?
Of that 141 seconds, we spend nearly 66 seconds just generating typecaches.
Turns out the storage component is monumentally stupid, and uses string lists to cache a typecache of the list passed into it, so we save on memory by only storing one list, but we lose on cpu time because we need to generate a new typecache for each list, and god trust me these lists hardly vary.

 So anyway that's like 50% of our worst case gone https://github.com/LemonInTheDark/tgstation/commit/9e197b584e6a2631cf7f55eb97309adca8d00ceb
Moving on.

The main reason we need to do so much tine on preferences is because every change to the prefs menu triggers a full rebuild of the dummy.
This is to make preference options development as simple as possible, but it places some pretty significant overhead on reloading, since it turns out spawning in humans is slow.

But why exactly is it slow?
Well there's 2 main painpoints. Painpoint 1 is generating organs. Each time you reload the prefs menu, we need to regenerate your species from a template. This means deleting and respawning your organs.

Painpoint 2 is equipping the mob. Typecaches aside, generating the equipment needed to kit out any reasonably advanced job like paramedic or engineer takes a pretty significant chunk of time.

Starting with organs, we only really need to spawn in organs that you can like, see, right? Like it's not like the dummies breath, or have blood or eat food, so we can just like, not give them stomachs and lungs and such.

That's what https://github.com/LemonInTheDark/tgstation/commit/b78352728a66448d7fcc5480c05d572f24402d26 did.
It didn't save as much time as I claimed in that commit desc, but it's still not a bad amount. 9 organs to ~3 is pretty damn good.

Ok so items. It's not like we can just not spawn them in, since we have no idea what humans need to see and what they don't, and modifying job code to suit that goal is not reasonable.

So instead, why not just, not spawn the items immediately? We could queue them up in a subsystem, and then just hand then out on request.
https://github.com/LemonInTheDark/tgstation/commit/a62b591d5b0f1ee172675571ec54c852b36ef62e

There's a whole lot more to this then just this commit, most of this pr is setup to make this system of object pooling as not bad as possible.
It's by design always going to be a bit fragile, since objects can do strange things on initialize, or have weird expectations.
But if it's used in a limited scope that's not the worst thing to deal with, and god it's so fast.

Ok so things are better, but we still spend a lot of time creating and deleting organs. Let's just queue those too
https://github.com/LemonInTheDark/tgstation/commit/f9d82df3454f702fac6d4887b8130c79550f214f
(I later walked back part of this, queuing limbs is kinda pointless and brains are just more trouble then they're worth)

That's even better, but our cache can fail if someone spams the menu, and we're still spending way more time then we should be deleting equipment and organs.
So why not just like, not? It's not like dummies are running around the map using their equipment, so when we delete it couldn't we insert it back into our queue? Same thing with the organs really, except they need special handling for extraction and insertion to disable/enable their rotting mechanic.

https://github.com/LemonInTheDark/tgstation/commit/c862a783899e3619ce58d16d16d13688eb2cefb2
This is really really fast. We obviously can't do this with everything, if we reinserted backpacks for instance we could end up duping their contents, which is A:weird for players, and B:a potential memory leak.
So let's pretend any of this is sane and just cache the things with no contents. If we want to speed up their generation, we can just cache the contents of the object, so loading it is faster.

Since we're doing this, let's cache the contents of belts, and also pdas. Unshits pda code slightly in the process.

Moving on, things are still a bit slow. In fact somehow ungeared runs are faster without caching then with it.
That's really weird. Most of the extra cost is in regenerate organs and harvest organs, and...
Oh.

Whenever you remove or insert an eye into someone, it calls regenerate_icons on the mob.
I can't even begin to tell you how slow this is, but to put it into perspective it takes roughly 10 seconds of the 40 we're spending loading cached ces.

Allllright, so we can actually just not, since like 4 years ago timegreen added the right proc, update_body (which updates eye color) to the bottom of the proc. 

https://github.com/LemonInTheDark/tgstation/commit/3173bca60d4af7c6308d90db008f78f140d378fc

Well that's much better, ahhhhhhh ok just about done then eh?
Wait a moment, why does my mob have no well, anything?
qc9ciZGV6k

Turns out that the regenerate_icons call in EYE REMOVAL IS THE ONLY REASON THAT DUMMIES RENDER LIMBS RIGHT NOW
https://github.com/LemonInTheDark/tgstation/commit/d817ef053875e38c9ec3552c3eea1856984298da

`update_body_parts` caches its overlay updates based on the makeup of its limbs, but since we empty the mob's overlay list as a part of dummy creation, this does nothing.

dependency

Alllright, finally done.

If you're curious, here are the final profiles
[Cached CE](https://github.com/tgstation/tgstation/files/7647247/cached.CE.csv)
[Cached, No Gear](https://github.com/tgstation/tgstation/files/7647248/cached.no.gear.csv)
[Uncached CE](https://github.com/tgstation/tgstation/files/7647249/uncached.CE.csv)
[Uncached, No Gear](https://github.com/tgstation/tgstation/files/7647250/uncached.no.gear.csv)

Caching saves 1.5x for renders with no gear, and 2x for renders with max gear. So pretty fucking good.

Looking at it from a before and after point of view, No gear rendering has gotten 3 times better, and Fully geared rendering has gotten 7 times faster.

Boooo yah suck it playerbase.

## Why It's Good For The Game

Papa needs some more execution time, OHHH fuck this whole thing just effects lag spikes doesn't it DAMN IT.
IDK maybe it makes the profiler easier?
Oh I guess you'll see less or hopefully no lag spikes at highpop when someone changes the color of their hair from #c034eb to #d934eb

If the lag spikes continue we can just move rerendering to a subsystem, but I'm not totally sure that's needed. We'll have to see.

One other thing, round start code makes significant use of preference dummies. Need to generate profiles for people's medical/security records and all.
I don't really have a good way of telling you how much what I'm doing here helps with that, since `get_flat_human_icon` also well, calls getFlatIcon, but it ought to at least help a bit.

Ah WAIT ONE MORE THING
Since object pooling and the optimizations done to human loading are generic, this speeds up at least somewhat the things that load humans. So roundstart again, just spawning in, etc.

TODO/Think about:
- [ ] Scale the amount of gear we preload to match pop (This is a maybe) 
- [x] Make a unit test that checks to make sure that everything loads proper, and that nothing qdel's itself or something. This way we can at least pretend that the system is robust 

TL;DR
Dummy go speed
Objects are cached and recached sometimes (sus)
Roundstart is maybe faster

## Changelog

:cl: LemonInTheDark
code: I've changed how job equipment and organs are loaded, if you end up with someone else's head/id card/eyes, please let me know yeah?
code: The preferences menu is now much more efficient. Hopefully you'll see less small jitters at highpop. 
/:cl:

### CODE REVIEW

File: code/game/objects/items/storage/belt.dm [hunk 1]
```diff
@@ -81,24 +81,62 @@
...
 	new /obj/item/analyzer/ranged(src)	//SKYRAT EDIT - original: new /obj/item/analyzer(src)
+=======
+	SSwardrobe.provide_type(/obj/item/screwdriver/power, src)
+	SSwardrobe.provide_type(/obj/item/crowbar/power, src)
+	SSwardrobe.provide_type(/obj/item/weldingtool/experimental, src)//This can be changed if this is too much //It's been 5 years
+	SSwardrobe.provide_type(/obj/item/multitool, src)
+	SSwardrobe.provide_type(/obj/item/stack/cable_coil, src)
+	SSwardrobe.provide_type(/obj/item/extinguisher/mini, src)
+	SSwardrobe.provide_type(/obj/item/analyzer, src)
+>>>>>>> 9b605b9cc0a (Speeds up the preference menu, significantly. Adds object pooling, other stuff too (#63225))
```
Comment (GoldenAlpharex) [inline code/game/objects/items/storage/belt.dm]:
```suggestion
	SSwardrobe.provide_type(/obj/item/screwdriver/power, src)
	SSwardrobe.provide_type(/obj/item/crowbar/power, src)
	SSwardrobe.provide_type(/obj/item/weldingtool/electric, src)//This can be changed if this is too much //It's been 5 years // SKYRAT EDIT - ORIGINAL: SSwardrobe.provide_type(/obj/item/weldingtool/experimental, src)
	SSwardrobe.provide_type(/obj/item/multitool, src)
	SSwardrobe.provide_type(/obj/item/stack/cable_coil, src)
	SSwardrobe.provide_type(/obj/item/extinguisher/mini, src)
	SSwardrobe.provide_type(/obj/item/analyzer/ranged, src) //SKYRAT EDIT - ORIGINAL: SSwardrobe.provide_type(/obj/item/analyzer, src)
```

File: code/game/objects/items/storage/belt.dm [hunk 2]
```diff
@@ -81,24 +81,62 @@
...
 
+/obj/item/storage/belt/utility/chief/full/get_types_to_preload()
+	var/list/to_preload = list() //Yes this is a pain. Yes this is the point
+	to_preload += /obj/item/screwdriver/power
+	to_preload += /obj/item/crowbar/power
+	to_preload += /obj/item/weldingtool/experimental
+	to_preload += /obj/item/multitool
+	to_preload += /obj/item/stack/cable_coil
+	to_preload += /obj/item/extinguisher/mini
+	to_preload += /obj/item/analyzer
```
Comment (GoldenAlpharex) [inline code/game/objects/items/storage/belt.dm]:
```suggestion
	to_preload += /obj/item/screwdriver/power
	to_preload += /obj/item/crowbar/power
	to_preload += /obj/item/weldingtool/electric // SKYRAT EDIT - Electric welder
	to_preload += /obj/item/multitool
	to_preload += /obj/item/stack/cable_coil
	to_preload += /obj/item/extinguisher/mini
	to_preload += /obj/item/analyzer/ranged // SKYRAT EDIT - Ranged Analyzer
```
